sdk-core/src/main/java/care/data4life/sdk/migration/RecordCompatibilityService.kt

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2021 D4L data4life gGmbH / All rights reserved.
+ *
+ * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
+ * including any intellectual property rights that subsist in the SDK.
+ *
+ * The SDK and its documentation may be accessed and used for viewing/review purposes only.
+ * Any usage of the SDK for other purposes, including usage for the development of
+ * applications/third-party applications shall require the conclusion of a license agreement
+ * between you and D4L.
+ *
+ * If you are interested in licensing the SDK for your own applications/third-party
+ * applications and/or if you’d like to contribute to the development of the SDK, please
+ * contact D4L by email to help@data4life.care.
+ */
+
+package care.data4life.sdk.migration
+
+import care.data4life.sdk.ApiService
+import care.data4life.sdk.network.model.NetworkModelContract.EncryptedRecord
+import care.data4life.sdk.tag.Annotations
+import care.data4life.sdk.tag.TaggingContract
+import care.data4life.sdk.tag.Tags
+import io.reactivex.Observable
+import io.reactivex.Single
+import io.reactivex.functions.BiFunction
+
+// see: https://gesundheitscloud.atlassian.net/browse/SDK-525
+class RecordCompatibilityService internal constructor(
+        private val apiService: ApiService,
+        private val tagEncryptionService: TaggingContract.EncryptionService
+) : MigrationContract.CompatibilityService {
+    private fun encrypt(
+            plainTags: Tags,
+            plainAnnotations: Annotations
+    ): Pair<List<String>, List<String>> {
+        return Pair(
+                tagEncryptionService.encryptAndEncodeTags(plainTags)
+                        .also { encryptedTags ->
+                            encryptedTags.addAll(
+                                    tagEncryptionService.encryptAndEncodeAnnotations(plainAnnotations)
+                            )
+                        },
+                tagEncryptionService.encryptTags(plainTags)
+                        .also { encryptedTags ->
+                            encryptedTags.addAll(
+                                    tagEncryptionService.encryptAnnotations(plainAnnotations)
+                            )
+                        }
+        )
+    }
+
+    private fun needsDoubleCall(
+            encodedAndEncryptedTags: List<String>,
+            encryptedTags: List<String>
+    ): Boolean = encodedAndEncryptedTags.sorted() != encryptedTags.sorted()
```

Comment:
Could you explain how that works? I don't get it.

Reply:
Also the migration logic bleeds into other parts of SDK -> TagEncryptionService. Ideally it is located in one place that could be easily removed when not needed anymore. The need for `@Deprecated` is also a strong sign for missing decoupling.

Reply:
> Could you explain how that works? I don't get it.
Sure!
The context can be found in [here](https://gesundheitscloud.atlassian.net/secure/RapidBoard.jspa?rapidView=109&projectKey=SDK&modal=detail&selectedIssue=SDK-534).

1.  the compatibility layer method gets a call
2. the method creats 2 version of the tags (one with url encoding one without)
3. then it checks if the 2 versions are equal (´needsDoubleCall`) since in case of arbitrary data we can encounter the case that both are equal (for example there no annotations or the annotations are [a-z]+)
We need this since otherwise we get the very same record twice.
The `sort` is arguably a little overdo in its given context now, but in my opinion is more 'safe', since context can change quickly.
4. the actual api call is executed
5. the results are concatenated

Reply:
> Also the migration logic bleeds into other parts of SDK -> TagEncryptionService, RecordService, ... Ideally it is located in one place that could be easily removed when not needed anymore. The need for @deprecated is also a strong sign for missing decoupling.

I see your point and I am open to suggestions.
In terms of the TagEncryptionService: The root cause for this is the coupling of the encoding and encrypting, which could be done separately. However the drawback is we loop twice over the tags then and in the compatibility context 3 times. My intension was to avoid that. What I also want to avoid is some sort of feature flag like iOS does it, since feature flags stay much longer than deprecated methods from my experience.
Also an alternative is to make `encryptList` public/internal, which is also not pretty from my perspective, and pull `encryptTags` out and place it here.

In terms of the RecordService: I found a bit odd to have encrypt and append for both in 2 places, but this is only a matter of taste. But what we can do is to introduce this encrypt&append to the TagEncryptionService and get rid of it entirely in RecordService. But this also depends on the way we deal with TagEncryptionService.

So what do you think?